webtransport/README.md

```diff
@@ -69,3 +69,39 @@ In order to verify end-to-end encryption of the connection, the peers need to es
 On receipt of the `webtransport_certhashes` extension, the client MUST verify that the certificate hash of the certificate that was used on the connection is contained in the server's list. If the client was willing to accept multiple certificate hashes, but cannot determine which certificate was actually used to establish the connection (this will commonly be the case for browser clients), it MUST verify that all certificate hashes are contained in the server's list. If verification fails, it MUST abort the handshake.
 
 For the client, the libp2p connection is fully established once it has sent the last Noise handshake message. For the server, processing of that message completes the handshake.
+
+## Message framing
+
+Upon closing a WebTransport stream, some implementations do not wait for all outstanding stream data to be sent over the wire before freeing up stream resources and making queue memory available for other uses.
+
+To allow cleanly closing streams without the loss of data using these implementations it's necessary to adopt similar semantics to the WebRTC transport when [closing datachannels](../webrtc/README.md#closing-an-rtcdatachannel).
+
+During the [security handshake](#security-handshake), a `webtransport_message_framing` boolean value may be supplied as part of the Noise extension block.
+
+If present, all messages sent over this connection will framed within a protobuf wrapper with the following structure:
+
+```proto
+syntax = "proto3";
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
```

Comment:
Not sure if this is an issue, but the `flag` field is optional and the `0` value starts closing the stream.

If you were to write a data message following `proto3` semantics which let you omit the `flag` field, then read the message using `proto2` semantics, the `flag` field would get initialised to it's default value of `0` and the receiver would think the sender wanted to close the stream.

We can work around this by having a `NIL` flag as the `0` value, which would be ignored by any receiver when read by a `proto2` or `proto3` decoder.

Or we expect implementations to apply `proto3` semantics because that's what the definition here says.